### PR TITLE
Refactor package structure and code, fix bugs

### PR DESCRIPTION
**Pull request by qpenko**
This commit focuses on further polishing the internal structure of the package, fixes bug, and adds some functionality.

- Two main modules (`matchblock.py` and `utils.py`) were renamed to emphasize their non-user accessible nature
- Package namespace only contains public class and functions, each module specifies its own set of names included in the `__all__` variables
- Tolerance interval checking decorator was refactored and the bug found there was fixed
- Tolerance class attributes implementation now uses descriptor protocol, making it better suited to the idea behind their role
- Optional parameters in the `MatchBlock` class add a new level of granularity to the initial data extraction process by allowing user to selectively turn off certain methods
- User accessible attributes of the `MatchBlock` class are implemented via property getters to accentuate that they are not expected to be updated manually
- `__repr__`, `__str__`, and `__eq__` methods of the `MatchBlock` were refactored
- Split methods were unified
- Other minor refactorings of the code and docstrings

### CODE REVIEW

File: docs/cookbook.rst [hunk 1]
```diff
@@ -207,10 +205,10 @@ integers. The next section includes such objects.
 In a real work situation you will probably want to perform more complex
 analysis. For example, you may want to determine whether a record from
 Database A is equal to a record from Database B. This can be achieved with
-match\_rows() function.
+match\_rows function.
```
Comment (dKiwi) [inline docs/cookbook.rst]: Function/method names below are in bold, perhaps we should be consistent?

**Comment by qpenko** [comment]
Thanks for spotting that one. While at it, I was also able to discover and fix other inconsistency and display issues with the documentation.

File: matchtools/_matchblock.py [hunk 2]
```diff
@@ -11,91 +11,187 @@
...
-    if len(dec_args) == 2 and callable(dec_args[1]):
-        # only lower bound provided
-        lower_bound, upper_bound = dec_args[0], None
-    elif len(dec_args) == 2:
-        # both bounds provided
-        lower_bound, upper_bound = dec_args[0], dec_args[1]
-    else:
-        raise ValueError('too many arguments: {}'.format(dec_args))
+    def __init__(self, attribute):
+        self.attribute = '_' + attribute
```
Comment (dKiwi) [inline matchtools/_matchblock.py:51]: I can see people use those operators in descriptors but perhaps we should consider join()?

**Review by dKiwi — COMMENTED** [review]
Good stuff. Just one comment regarding string concatenation in the descriptor. Thanks!

File: matchtools/_matchblock.py (hunk 2 above)
Comment (qpenko) [reply matchtools/_matchblock.py:51]: If your only concern regarding this is speed, I think it's not an issue here: this operation performed only once, and we are not joining tens or hundreds of strings together, where join() is the preferred way. In our case though, I think readability is better.

In fact, I have even made a quick measurement here, and the result is quite surprising to me:

```
In [1]: %timeit ''.join(('_', 'attribute'))
159 ns ± 3.33 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [2]: %timeit '_' + 'attribute'
17.2 ns ± 0.0439 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
```

So it turns out, that in current CPython implementation, it's actually faster to join two strings together using addition operator rather that string join() method.

Also, I like this short [answer](https://softwareengineering.stackexchange.com/a/304621) regarding string concatenation problem.

**Review by dKiwi — APPROVED** [review]

**Pull request merged**